[filebeat][streaming] - Added support for TLS & forward proxy configs for websockets

ShourieG opened the pull request:
## Type of change
- Enhancement
- Docs

## Proposed commit message
Added support for functional TLS config & proxy configs for websockets by implementing a custom transport object and NetDialContext() and leveraging the httpcommon library. Added supporting tests to accompany

## Note: 
Tried using in memory certs and keys but our current httpcommon library always asks for a file.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact
No enduser impact as this is all additive.
<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Closes 
- Relates https://github.com/elastic/integrations/issues/11581

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

ShourieG commented on x-pack/filebeat/docs/inputs/input-streaming.asciidoc:
```diff
@@ -349,6 +349,32 @@ The minimum time to wait between retries. This ensures that retries are spaced o
 
+[float]
+==== `handshake_timeout`
+This specifies the time to wait for the `websocket` handshake and the `http.Upgrade()` operation to complete. This timeout occurs at the application layer and not at the TCP layer. The `default value` is `20` seconds. This setting is specific to the `websocket` streaming input type only.
```
@efd6,
What I was going for initially was faster failures at the application layer to achieve better program responsiveness. But then I realised that we might want to wait for TCP retires to overcome any intermittent network issues present and setting a handshake timeout lesser than the tcp timeout defeats that purpose. 

The **timeout option below** actually controls the net.Dialer.DialContext timeout, which by default is set to 90s, half of the os default for TCP timeout window.

So my question is, should we extend the  DialContext and handshake timeouts to mirror the TCP time out value of the os ? 

As for the documentation, I'll update it.

efd6 replied:
I suggest taking a look at the gorilla code to see how the timeout is propagated.

ShourieG replied:
@efd6, I saw the gorilla code just creates a context deadline out of the provided timeout in the dialcontext. So in this case having the handshake value < the dial context doesn't make much sense by default. We can keep them at the same value then if the user feels they need to change the can do that.

efd6 replied:
I would not bother adding the handshake timeout since we already provide a mechanism to hand in a dial timeout which achieves the same end.

ShourieG commented:
@efd6, I've addressed the PR suggestions. I was wondering should we back port this to 8.17 ? I know the FF has already happened but this might help out a lot of users currently facing issues.
